File: .agents/testing/pi_frontend_test.py
#!/usr/bin/env python3
"""
Focused Playwright tests for Pi Chat frontend.
Tests the actual single-page chat application (no auth, no routing).
"""

import sys
import os

# Add playwright to path
sys.path.insert(0, '/home/jon/git/betty/.playwright-cli')

from playwright.sync_api import sync_playwright
import time

def run_tests():
    errors = []
    warnings = []
    
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(viewport={'width': 1280, 'height': 720})
        page = context.new_page()
        
        # Track console errors
        console_errors = []
        page.on('pageerror', lambda error: console_errors.append(str(error)))
        page.on('console', lambda msg: print(f"  Console [{msg.type}]: {msg.text}") if msg.type in ('error', 'warning') else None)
        
        # ============================================================
        # TEST 1: App loads without JS errors
        # ============================================================
        print("=" * 60)
        print("TEST 1: App loads without JS errors")
        print("=" * 60)
        try:
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            
            # Check the page title
            title = page.title()
            print(f"  Page title: {title}")
            if title == "Pi Chat":
                print("  PASS: Correct title")
            else:
                warnings.append(f"Unexpected title: {title}")
            
            # Check for the app container
            app = page.query_selector('.app')
            if app:
                print("  PASS: App container found")
            else:
                errors.append("App container (.app) not found")
            
            # Check for header
            header = page.query_selector('.app-header')
            if header:
                print("  PASS: Header found")
            else:
                errors.append("Header (.app-header) not found")
            
            # Check for chat container
            chat_container = page.query_selector('.chat-container')
            if chat_container:
                print("  PASS: Chat container found")
            else:
                errors.append("Chat container (.chat-container) not found")
            
            # Check for input area
            input_area = page.query_selector('.input-area')
            if input_area:
                print("  PASS: Input area found")
            else:
                errors.append("Input area (.input-area) not found")
                
        except Exception as e:
            errors.append(f"App load error: {e}")
        
        # ============================================================
        # TEST 2: Welcome screen is displayed
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 2: Welcome screen")
        print("=" * 60)
        try:
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(1)  # Wait for Vue to render
            
            welcome = page.query_selector('.welcome-screen')
            if welcome:
                print("  PASS: Welcome screen visible")
                
                # Check for welcome heading
                h2 = welcome.query_selector('h2')
                if h2:
                    text = h2.inner_text().strip()
                    print(f"  Welcome heading: {text}")
                    if "Pi Chat" in text:
                        print("  PASS: Correct welcome heading")
                    else:
                        warnings.append(f"Unexpected welcome heading: {text}")
                
                # Check for welcome hints
                hints = welcome.query_selector_all('.welcome-hint')
                print(f"  Welcome hints: {len(hints)}")
                if len(hints) > 0:
                    print("  PASS: Welcome hints present")
                    for i, hint in enumerate(hints[:2]):
                        print(f"    Hint {i+1}: {hint.inner_text().strip()[:60]}")
                else:
                    errors.append("No welcome hints found")
            else:
                errors.append("Welcome screen not found")
                
        except Exception as e:
            errors.append(f"Welcome screen error: {e}")
        
        # ============================================================
        # TEST 3: Status badge shows correct state
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 3: Status badge")
        print("=" * 60)
        try:
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(2)  # Wait for WebSocket connection attempt
            
            status_badge = page.query_selector('.status-badge')
            if status_badge:
                print("  PASS: Status badge found")
                
                # Check for connected state (backend is running)
                if status_badge.get_attribute('class'):
                    classes = status_badge.get_attribute('class')
                    print(f"  Status classes: {classes}")
                    
                    if 'connected' in classes:
                        print("  PASS: Connected status shown")
                    elif 'connecting' in classes:
                        print("  WARN: Still connecting (may be slow WebSocket)")
                    elif 'error' in classes:
                        print("  WARN: Error status shown")
                        warnings.append("Status badge shows error state")
                    
                    # Check status text
                    status_text = status_badge.inner_text().strip()
                    print(f"  Status text: {status_text}")
            else:
                errors.append("Status badge not found")
                
        except Exception as e:
            errors.append(f"Status badge error: {e}")
        
        # ============================================================
        # TEST 4: New Chat button exists and is enabled
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 4: New Chat button")
        print("=" * 60)
        try:
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(2)
            
            new_chat_btn = page.query_selector('.app-header .btn')
            if new_chat_btn:
                text = new_chat_btn.inner_text().strip()
                print(f"  Button text: '{text}'")
                disabled = new_chat_btn.get_attribute('disabled')
                print(f"  Disabled: {disabled}")
                if disabled:
                    print("  WARN: New Chat button is disabled (may need connection)")
                else:
                    print("  PASS: New Chat button enabled")
            else:
                errors.append("New Chat button not found")
                
        except Exception as e:
            errors.append(f"New Chat button error: {e}")
        
        # ============================================================
        # TEST 5: Input textarea exists and is functional
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 5: Input textarea")
        print("=" * 60)
        try:
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(2)
            
            textarea = page.query_selector('textarea')
            if textarea:
                print("  PASS: Textarea found")
                placeholder = textarea.get_attribute('placeholder')
                print(f"  Placeholder: {placeholder}")
                disabled = textarea.get_attribute('disabled')
                print(f"  Disabled: {disabled}")
                
                # Test typing
                textarea.fill("Hello, Pi!")
                value = textarea.input_value()
                print(f"  After typing: '{value}'")
                if value == "Hello, Pi!":
                    print("  PASS: Textarea accepts input")
                else:
                    errors.append("Textarea value mismatch")
            else:
                errors.append("Textarea not found")
                
        except Exception as e:
            errors.append(f"Textarea error: {e}")
        
        # ============================================================
        # TEST 6: Send button exists
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 6: Send button")
        print("=" * 60)
        try:
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(2)
            
            send_btn = page.query_selector('.btn-primary')
            if send_btn:
                print("  PASS: Send button found")
                disabled = send_btn.get_attribute('disabled')
                print(f"  Disabled: {disabled}")
            else:
                errors.append("Send button not found")
                
        except Exception as e:
            errors.append(f"Send button error: {e}")
        
        # ============================================================
        # TEST 7: Format message rendering (via v-html)
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 7: Message formatting (formatMessage)")
        print("=" * 60)
        try:
            # Test formatMessage function via page.evaluate
            result = page.evaluate('''() => {
                // Replicate the formatMessage function from App.vue
                function formatMessage(content) {
                    if (!content) return "";
                    let text = content
                        .replace(/&/g, "&amp;")
                        .replace(/</g, "&lt;")
                        .replace(/>/g, "&gt;");
                    text = text.replace(/```(\\w*)\\n([\\s\\S]*?)```/g, (_, lang, code) => {
                        return `<pre><code class="language-${lang}">${code.trim()}</code></pre>`;
                    });
                    text = text.replace(/`([^`]+)`/g, '<code>$1</code>');
                    text = text.replace(/\\*\\*([\\s\\S]*?)\\*\\*/g, "<strong>$1</strong>");
                    text = text.replace(/\\*([\\s\\S]*?)\\*/g, "<em>$1</em>");
                    text = text.replace(/(<pre>.*?<\\/pre>)|(\\n)/gs, (match, preBlock) => {
                        return preBlock !== undefined ? preBlock : "<br>";
                    });
                    return text;
                }
                return {
                    plain: formatMessage("Hello world"),
                    bold: formatMessage("**bold text**"),
                    italic: formatMessage("*italic text*"),
                    code: formatMessage("`inline code`"),
                    multiline: formatMessage("Line 1\nLine 2\nLine 3"),
                    xss: formatMessage("<script>alert('xss')</script>"),
                    codeblock: formatMessage("```js\nconsole.log('hi')\n```"),
                };
            }''')
            
            print(f"  Plain text: {result['plain'][:50]}...")
            print(f"  Bold: {result['bold']}")
            print(f"  Italic: {result['italic']}")
            print(f"  Code: {result['code']}")
            print(f"  Multiline: {result['multiline'][:50]}...")
            print(f"  XSS test: {result['xss'][:50]}...")
            print(f"  Code block: {result['codeblock'][:60]}...")
            
            # Check XSS escaping
            if '<script>' not in result['xss']:
                print("  PASS: XSS content properly escaped")
            else:
                errors.append("XSS content NOT escaped!")
            
            # Check bold/italic rendering
            if '<strong>' in result['bold']:
                print("  PASS: Bold text rendered")
            else:
                errors.append("Bold text not rendered")
            
            if '<em>' in result['italic']:
                print("  PASS: Italic text rendered")
            else:
                errors.append("Italic text not rendered")
            
            # Check newlines in multiline
            if '<br>' in result['multiline']:
                print("  PASS: Line breaks rendered")
            else:
                errors.append("Line breaks not rendered")
            
            # Check code block
            if '<pre>' in result['codeblock']:
                print("  PASS: Code block rendered")
            else:
                errors.append("Code block not rendered")
                
        except Exception as e:
            errors.append(f"Format message error: {e}")
            import traceback
            print(f"  Error: {e}")
            traceback.print_exc()
        
        # ============================================================
        # TEST 8: Responsive design - mobile view (375px)
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 8: Mobile responsive (375px)")
        print("=" * 60)
        try:
            page.set_viewport_size({'width': 375, 'height': 667})
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(1)
            
            # Check that content is visible
            app = page.query_selector('.app')
            if app:
                bbox = app.bounding_box()
                print(f"  App visible: {bbox is not None}")
                if bbox:
                    print(f"  App width: {bbox['width']}, height: {bbox['height']}")
                    print("  PASS: App renders at mobile width")
            else:
                errors.append("App not visible at mobile width")
                
        except Exception as e:
            errors.append(f"Mobile responsive error: {e}")
        
        # ============================================================
        # TEST 9: Responsive design - tablet view (768px)
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 9: Tablet responsive (768px)")
        print("=" * 60)
        try:
            page.set_viewport_size({'width': 768, 'height': 1024})
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(1)
            
            app = page.query_selector('.app')
            if app:
                print("  PASS: App renders at tablet width")
            else:
                errors.append("App not visible at tablet width")
                
        except Exception as e:
            errors.append(f"Tablet responsive error: {e}")
        
        # ============================================================
        # TEST 10: Responsive design - desktop view (1920px)
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 10: Desktop responsive (1920px)")
        print("=" * 60)
        try:
            page.set_viewport_size({'width': 1920, 'height': 1080})
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(1)
            
            app = page.query_selector('.app')
            if app:
                print("  PASS: App renders at desktop width")
            else:
                errors.append("App not visible at desktop width")
                
        except Exception as e:
            errors.append(f"Desktop responsive error: {e}")
        
        # ============================================================
        # TEST 11: WebSocket connection to backend
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 11: WebSocket connection")
        print("=" * 60)
        try:
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(3)  # Wait for WebSocket connection
            
            # Check WebSocket status via JS
            ws_status = page.evaluate('''() => {
                return new Promise((resolve) => {
                    // Check if there's a WebSocket connection
                    const openSockets = [];
                    // We can't directly check WebSocket state from here,
                    // but we can check the Vue app state
                    const app = document.querySelector('.status-badge');
                    if (app) {
                        resolve({
                            hasBadge: true,
                            classes: app.getAttribute('class'),
                            text: app.innerText.trim()
                        });
                    } else {
                        resolve({ hasBadge: false });
                    }
                });
            }''')
            
            print(f"  WebSocket status: {ws_status}")
            if ws_status.get('hasBadge'):
                if 'connected' in ws_status.get('classes', ''):
                    print("  PASS: WebSocket connected to backend")
                elif 'error' in ws_status.get('classes', ''):
                    errors.append("WebSocket connection failed")
                else:
                    warnings.append(f"WebSocket status unclear: {ws_status}")
            else:
                errors.append("Status badge not found")
                
        except Exception as e:
            errors.append(f"WebSocket connection error: {e}")
        
        # ============================================================
        # TEST 12: Error handling - offline state
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 12: Error banner display")
        print("=" * 60)
        try:
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(2)
            
            # Check if error banner is visible (shouldn't be if connected)
            error_banner = page.query_selector('.error-banner')
            if error_banner:
                print("  WARN: Error banner is visible")
                error_text = error_banner.inner_text().strip()
                print(f"  Error text: {error_text}")
            else:
                print("  PASS: No error banner (connection successful)")
                
        except Exception as e:
            errors.append(f"Error banner test error: {e}")
        
        # ============================================================
        # TEST 13: Console errors check
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 13: Console errors")
        print("=" * 60)
        try:
            if console_errors:
                print(f"  {len(console_errors)} console error(s) found:")
                for err in console_errors:
                    print(f"    ✗ {err[:200]}")
                errors.append(f"{len(console_errors)} console error(s) detected")
            else:
                print("  PASS: No console errors")
        except Exception as e:
            errors.append(f"Console errors check error: {e}")
        
        # ============================================================
        # TEST 14: New session flow
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 14: New session flow")
        print("=" * 60)
        try:
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(2)
            
            # Check that welcome screen is visible (no messages yet)
            messages = page.query_selector_all('.message')
            print(f"  Initial messages: {len(messages)}")
            
            # Click a welcome hint to start a conversation
            hints = page.query_selector_all('.welcome-hint')
            if hints:
                print(f"  Clicking welcome hint...")
                hints[0].click()
                time.sleep(2)
                
                # Check that user message appeared
                messages = page.query_selector_all('.message')
                print(f"  Messages after click: {len(messages)}")
                if len(messages) > 0:
                    print("  PASS: User message appeared after clicking hint")
                else:
                    warnings.append("No messages after clicking hint (WebSocket may not have connected)")
                
                # Check that streaming indicator appears
                streaming = page.query_selector('.message.assistant')
                if streaming:
                    print("  PASS: Streaming message indicator visible")
                else:
                    warnings.append("No streaming message visible (expected - Pi may not respond)")
            else:
                errors.append("No welcome hints found for session test")
                
        except Exception as e:
            errors.append(f"New session flow error: {e}")
        
        # ============================================================
        # TEST 15: Take screenshots
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST 15: Screenshots")
        print("=" * 60)
        try:
            page.set_viewport_size({'width': 1280, 'height': 720})
            page.goto('http://127.0.0.1:5173', timeout=15000)
            page.wait_for_load_state('networkidle', timeout=10000)
            time.sleep(2)
            page.screenshot(path='/home/jon/git/betty/.agents/testing/SNAPSHOTS/welcome_screen.png')
            print("  Screenshot saved: welcome_screen.png")
            
            # Mobile screenshot
            page.set_viewport_size({'width': 375, 'height': 667})
            page.screenshot(path='/home/jon/git/betty/.agents/testing/SNAPSHOTS/mobile_view.png')
            print("  Screenshot saved: mobile_view.png")
                
        except Exception as e:
            errors.append(f"Screenshot error: {e}")
        
        # ============================================================
        # Final Summary
        # ============================================================
        print("\n" + "=" * 60)
        print("TEST SUMMARY")
        print("=" * 60)
        print(f"  Errors: {len(errors)}")
        print(f"  Warnings: {len(warnings)}")
        for err in errors:
            print(f"    ✗ {err}")
        for warn in warnings:
            print(f"    ⚠ {warn}")
        
        browser.close()
        return len(errors)

if __name__ == '__main__':
    error_count = run_tests()
    sys.exit(1 if error_count > 0 else 0)

File: .agents/testing/pi_frontend_test.mjs
import { chromium } from "playwright";

async function runTests() {
  const errors = [];
  const warnings = [];
  let browser;

  try {
    browser = await chromium.launch({ headless: true });
    const context = await browser.newContext({ viewport: { width: 1280, height: 720 } });
    const page = await context.newPage();

    // Track console errors
    const consoleErrors = [];
    page.on("pageerror", (error) => consoleErrors.push(error.message));
    page.on("console", (msg) => {
      if (msg.type() === "error" || msg.type() === "warning") {
        console.log(`  Console [${msg.type()}]: ${msg.text()}`);
      }
    });

    // ============================================================
    // TEST 1: App loads without JS errors
    // ============================================================
    console.log("=" .repeat(60));
    console.log("TEST 1: App loads without JS errors");
    console.log("=" .repeat(60));

    try {
      await page.goto("http://127.0.0.1:5173", { waitUntil: "networkidle", timeout: 15000 });
      await page.waitForLoadState("networkidle");

      const title = await page.title();
      console.log(`  Page title: ${title}`);
      if (title === "Pi Chat") {
        console.log("  PASS: Correct title");
      } else {
        warnings.push(`Unexpected title: ${title}`);
      }

      const app = await page.$(".app");
      if (app) {
        console.log("  PASS: App container found");
      } else {
        errors.push("App container (.app) not found");
      }

      const header = await page.$(".app-header");
      if (header) {
        console.log("  PASS: Header found");
      } else {
        errors.push("Header (.app-header) not found");
      }

      const chatContainer = await page.$(".chat-container");
      if (chatContainer) {
        console.log("  PASS: Chat container found");
      } else {
        errors.push("Chat container (.chat-container) not found");
      }

      const inputArea = await page.$(".input-area");
      if (inputArea) {
        console.log("  PASS: Input area found");
      } else {
        errors.push("Input area (.input-area) not found");
      }
    } catch (e) {
      errors.push(`App load error: ${e.message}`);
    }

    // ============================================================
    // TEST 2: Welcome screen is displayed
    // ============================================================
    console.log("\n" + "=".repeat(60));
    console.log("TEST 2: Welcome screen");
    console.log("=" .repeat(60));

    try {
      await page.goto("http://127.0.0.1:5173", { waitUntil: "networkidle", timeout: 15000 });
      await page.waitForLoadState("networkidle");
      await page.waitForTimeout(1500);

      const welcome = await page.$(".welcome-screen");
      if (welcome) {
        console.log("  PASS: Welcome screen visible");

        const h2 = await welcome.$("h2");
        if (h2) {
          const text = await h2.innerText();
          console.log(`  Welcome heading: ${text}`);
          if (text.includes("Pi Chat")) {
            console.log("  PASS: Correct welcome heading");
          } else {
            warnings.push(`Unexpected welcome heading: ${text}`);
          }
        }

        const hints = await welcome.$$(".welcome-hint");
        console.log(`  Welcome hints: ${hints.length}`);
        if (hints.length > 0) {
          console.log("  PASS: Welcome hints present");
          for (let i = 0; i < Math.min(2, hints.length); i++) {
            const hintText = await hints[i].innerText();
            console.log(`    Hint ${i + 1}: ${hintText.substring(0, 60)}`);
          }
        } else {
          errors.push("No welcome hints found");
        }
      } else {
        errors.push("Welcome screen not found");
      }
    } catch (e) {
      errors.push(`Welcome screen error: ${e.message}`);
    }

    // ============================================================
    // TEST 3: Status badge shows correct state
    // ============================================================
    console.log("\n" + "=".repeat(60));
    console.log("TEST 3: Status badge");
    console.log("=" .repeat(60));

    try {
      await page.goto("http://127.0.0.1:5173", { waitUntil: "networkidle", timeout: 15000 });
      await page.waitForLoadState("networkidle");
      await page.waitForTimeout(3000);

      const statusBadge = await page.$(".status-badge");
      if (statusBadge) {
        console.log("  PASS: Status badge found");
        const classes = await statusBadge.getAttribute("class");
        const text = await statusBadge.innerText();
        console.log(`  Status classes: ${classes}`);
        console.log(`  Status text: ${text}`);

        if (classes.includes("connected")) {
          console.log("  PASS: Connected status shown");
        } else if (classes.includes("connecting")) {
          console.log("  WARN: Still connecting (may be slow WebSocket)");
        } else if (classes.includes("error")) {
          console.log("  WARN: Error status shown");
          warnings.push("Status badge shows error state");
        }
      } else {
        errors.push("Status badge not found");
      }
    } catch (e) {
      errors.push(`Status badge error: ${e.message}`);
    }

    // ============================================================
    // TEST 4: Message formatting (formatMessage function)
    // ============================================================
    console.log("\n" + "=".repeat(60));
    console.log("TEST 4: Message formatting (formatMessage)");
    console.log("=" .repeat(60));

    try {
      const result = await page.evaluate(() => {
        function formatMessage(content) {
          if (!content) return "";
          let text = content
            .replace(/&/g, "&amp;")
            .replace(/</g, "&lt;")
            .replace(/>/g, "&gt;");
          text = text.replace(/```(\w*)\n([\s\S]*?)```/g, (_, lang, code) => {
            return `<pre><code class="language-${lang}">${code.trim()}</code></pre>`;
          });
          text = text.replace(/`([^`]+)`/g, '<code>$1</code>');
          text = text.replace(/\*\*([\s\S]*?)\*\*/g, "<strong>$1</strong>");
          text = text.replace(/\*([\s\S]*?)\*/g, "<em>$1</em>");
          text = text.replace(/(<pre>.*?<\/pre>)|(\n)/gs, (match, preBlock) => {
            return preBlock !== undefined ? preBlock : "<br>";
          });
          return text;
        }
        return {
          plain: formatMessage("Hello world"),
          bold: formatMessage("**bold text**"),
          italic: formatMessage("*italic text*"),
          code: formatMessage("`inline code`"),
          multiline: formatMessage("Line 1\nLine 2\nLine 3"),
          xss: formatMessage("<script>alert('xss')</script>"),
          codeblock: formatMessage("```js\nconsole.log('hi')\n```"),
        };
      });

      console.log(`  Plain text: ${result.plain.substring(0, 50)}...`);
      console.log(`  Bold: ${result.bold}`);
      console.log(`  Italic: ${result.italic}`);
      console.log(`  Code: ${result.code}`);
      console.log(`  Multiline: ${result.multiline.substring(0, 50)}...`);
      console.log(`  XSS test: ${result.xss.substring(0, 50)}...`);
      console.log(`  Code block: ${result.codeblock.substring(0, 60)}...`);

      if (!result.xss.includes("<script>")) {
        console.log("  PASS: XSS content properly escaped");
      } else {
        errors.push("XSS content NOT escaped!");
      }

      if (result.bold.includes("<strong>")) {
        console.log("  PASS: Bold text rendered");
      } else {
        errors.push("Bold text not rendered");
      }

      if (result.italic.includes("<em>")) {
        console.log("  PASS: Italic text rendered");
      } else {
        errors.push("Italic text not rendered");
      }

      if (result.multiline.includes("<br>")) {
        console.log("  PASS: Line breaks rendered");
      } else {
        errors.push("Line breaks not rendered");
      }

      if (result.codeblock.includes("<pre>")) {
        console.log("  PASS: Code block rendered");
      } else {
        errors.push("Code block not rendered");
      }
    } catch (e) {
      errors.push(`Format message error: ${e.message}`);
      console.log(`  Error: ${e.message}`);
    }

    // ============================================================
    // TEST 5: Responsive design - mobile (375px)
    // ============================================================
    console.log("\n" + "=".repeat(60));
    console.log("TEST 5: Mobile responsive (375px)");
    console.log("=" .repeat(60));

    try {
      await page.setViewportSize({ width: 375, height: 667 });
      await page.goto("http://127.0.0.1:5173", { waitUntil: "networkidle", timeout: 15000 });
      await page.waitForLoadState("networkidle");
      await page.waitForTimeout(1000);

      const app = await page.$(".app");
      if (app) {
        const bbox = await app.boundingBox();
        console.log(`  App visible: ${bbox !== null}`);
        if (bbox) {
          console.log(`  App width: ${bbox.width}, height: ${bbox.height}`);
          console.log("  PASS: App renders at mobile width");
        }
      } else {
        errors.push("App not visible at mobile width");
      }
    } catch (e) {
      errors.push(`Mobile responsive error: ${e.message}`);
    }

    // ============================================================
    // TEST 6: Responsive design - tablet (768px)
    // ============================================================
    console.log("\n" + "=".repeat(60));
    console.log("TEST 6: Tablet responsive (768px)");
    console.log("=" .repeat(60));

    try {
      await page.setViewportSize({ width: 768, height: 1024 });
      await page.goto("http://127.0.0.1:5173", { waitUntil: "networkidle", timeout: 15000 });
      await page.waitForLoadState("networkidle");
      await page.waitForTimeout(1000);

      const app = await page.$(".app");
      if (app) {
        console.log("  PASS: App renders at tablet width");
      } else {
        errors.push("App not visible at tablet width");
      }
    } catch (e) {
      errors.push(`Tablet responsive error: ${e.message}`);
    }

    // ============================================================
    // TEST 7: Responsive design - desktop (1920px)
    // ============================================================
    console.log("\n" + "=".repeat(60));
    console.log("TEST 7: Desktop responsive (1920px)");
    console.log("=" .repeat(60));

    try {
      await page.setViewportSize({ width: 1920, height: 1080 });
      await page.goto("http://127.0.0.1:5173", { waitUntil: "networkidle", timeout: 15000 });
      await page.waitForLoadState("networkidle");
      await page.waitForTimeout(1000);

      const app = await page.$(".app");
      if (app) {
        console.log("  PASS: App renders at desktop width");
      } else {
        errors.push("App not visible at desktop width");
      }
    } catch (e) {
      errors.push(`Desktop responsive error: ${e.message}`);
    }

    // ============================================================
    // TEST 8: Input functionality
    // ============================================================
    console.log("\n" + "=".repeat(60));
    console.log("TEST 8: Input textarea");
    console.log("=" .repeat(60));

    try {
      await page.setViewportSize({ width: 1280, height: 720 });
      await page.goto("http://127.0.0.1:5173", { waitUntil: "networkidle", timeout: 15000 });
      await page.waitForLoadState("networkidle");
      await page.waitForTimeout(2000);

      const textarea = await page.$("textarea");
      if (textarea) {
        console.log("  PASS: Textarea found");
        const placeholder = await textarea.getAttribute("placeholder");
        console.log(`  Placeholder: ${placeholder}`);

        await textarea.fill("Hello, Pi!");
        const value = await textarea.inputValue();
        console.log(`  After typing: '${value}'`);
        if (value === "Hello, Pi!") {
          console.log("  PASS: Textarea accepts input");
        } else {
          errors.push("Textarea value mismatch");
        }
      } else {
        errors.push("Textarea not found");
      }
    } catch (e) {
      errors.push(`Textarea error: ${e.message}`);
    }

    // ============================================================
    // TEST 9: New session flow (click welcome hint)
    // ============================================================
    console.log("\n" + "=".repeat(60));
    console.log("TEST 9: New session flow (click welcome hint)");
    console.log("=" .repeat(60));

    try {
      await page.goto("http://127.0.0.1:5173", { waitUntil: "networkidle", timeout: 15000 });
      await page.waitForLoadState("networkidle");
      await page.waitForTimeout(2000);

      const hints = await page.$$(".welcome-hint");
      if (hints.length > 0) {
        console.log(`  Found ${hints.length} welcome hints`);
        await hints[0].click();
        await page.waitForTimeout(2000);

        const messages = await page.$$(".message");
        console.log(`  Messages after click: ${messages.length}`);
        if (messages.length > 0) {
          console.log("  PASS: User message appeared after clicking hint");
        } else {
          warnings.push("No messages after clicking hint (WebSocket may not have connected)");
        }

        const streaming = await page.$(".message.assistant");
        if (streaming) {
          console.log("  PASS: Streaming message indicator visible");
        } else {
          console.log("  INFO: No streaming message (expected - Pi may not respond)");
        }
      } else {
        errors.push("No welcome hints found for session test");
      }
    } catch (e) {
      errors.push(`New session flow error: ${e.message}`);
    }

    // ============================================================
    // TEST 10: Screenshot
    // ============================================================
    console.log("\n" + "=".repeat(60));
    console.log("TEST 10: Screenshots");
    console.log("=" .repeat(60));

    try {
      await page.setViewportSize({ width: 1280, height: 720 });
      await page.goto("http://127.0.0.1:5173", { waitUntil: "networkidle", timeout: 15000 });
      await page.waitForLoadState("networkidle");
      await page.waitForTimeout(2000);
      await page.screenshot({ path: "/home/jon/git/betty/.agents/testing/SNAPSHOTS/welcome_screen.png" });
      console.log("  Screenshot saved: welcome_screen.png");

      await page.setViewportSize({ width: 375, height: 667 });
      await page.screenshot({ path: "/home/jon/git/betty/.agents/testing/SNAPSHOTS/mobile_view.png" });
      console.log("  Screenshot saved: mobile_view.png");
    } catch (e) {
      errors.push(`Screenshot error: ${e.message}`);
    }

    // ============================================================
    // Console errors summary
    // ============================================================
    console.log("\n" + "=".repeat(60));
    console.log("CONSOLE ERRORS SUMMARY");
    console.log("=" .repeat(60));

    if (consoleErrors.length > 0) {
      console.log(`  ${consoleErrors.length} console error(s) found:`);
      for (const err of consoleErrors.slice(0, 10)) {
        console.log(`    ✗ ${err.substring(0, 200)}`);
      }
      errors.push(`${consoleErrors.length} console error(s) detected`);
    } else {
      console.log("  PASS: No console errors");
    }

    // ============================================================
    // Final Summary
    // ============================================================
    console.log("\n" + "=".repeat(60));
    console.log("TEST SUMMARY");
    console.log("=" .repeat(60));
    console.log(`  Errors: ${errors.length}`);
    console.log(`  Warnings: ${warnings.length}`);
    for (const err of errors) {
      console.log(`    ✗ ${err}`);
    }
    for (const warn of warnings) {
      console.log(`    ⚠ ${warn}`);
    }

    await browser.close();
    return errors.length;
  } catch (e) {
    console.error(`Fatal error: ${e.message}`);
    if (browser) await browser.close();
    return 1;
  }
}

runTests().then((errorCount) => {
  process.exit(errorCount > 0 ? 1 : 0);
});
